Test

anametiu opened the pull request.

anametiu committed:
Problem1 red

anametiu committed:
Problem1 red

anametiu committed:
Problem1 green

anametiu committed:
Problem1 refactor

anametiu committed:
Problem2 red

anametiu committed:
Problem2 green

anametiu committed:
Problem2 refactor

anametiu committed:
Problem3 red

anametiu committed:
Problem3 green

anametiu committed:
Problem3 refactor

drqCode commented on Problem1.rb:
```diff
@@ -0,0 +1,18 @@
+def  sum_even_numbers(array)
+
+  sum = 0
+  array.each { |x| sum += x.even? ? x : 0 }
```
Can you make it one line (without any aux variable)?

anametiu replied:
Yes: array.select(&:even?).sum
Commit: d2d8a7489a6358e37802db6ee26054408c72efad

drqCode commented on Problem2.rb:
```diff
@@ -0,0 +1,32 @@
+    puts expect(find_max_numbers([4, 8, 2, 6, 9, 3])).to eq([9, 8])
+    puts expect(find_max_numbers([1, 2, 3, 4, 5, 6])).to eq([6, 5])
+    puts expect(find_max_numbers([3, 2, 7, 1, 1, 1])).to eq([7, 3])
+
```
What if the array contains negative numbers?

alexandru-calinoiu replied:
Indeed integer numbers can be negative.

anametiu replied:
True, I initialized the first and second variables with minus infinite. Commit: 3664186d8d12b3b17bfeb43baaa175931c2f386b

drqCode commented on Problem3.rb:
```diff
@@ -0,0 +1,48 @@
+    temp = array[i]
+    j = temp - min
+
+    while i != j do
```
"Do not use 2 nesting for loops since the time complexity would be high. Try to find a solution by executing just one loop". Consider using just one loop, not matter if there is a very small one. :)

anametiu replied:
I checked again the implementation and you're right. Also the method swap extracted for refactoring makes the algorithm to fail for certain tests. I completely re implemented the requirement and committed in 79f5fe7284839366e7dcbe32786b4a2bb38b3243

alexandru-calinoiu commented on Problem3.rb:
```diff
@@ -0,0 +1,48 @@
+
+  it 'should find duplicates' do
+
+    puts contains_duplicates?([1, 2, 3, 4, 2])
```
Tests should contain asserts

anametiu replied:
I uploaded tests for problem1 and problem2 using assert

anametiu replied:
I implemented a different algorithm in commit 79f5fe7284839366e7dcbe32786b4a2bb38b3243

alexandru-calinoiu reviewed (changes requested):
Very good try, but problem 3 does not respect the requirements, you are suppose to solve it with only one loop.

pirvudoru commented on Problem3.rb:
```diff
@@ -0,0 +1,48 @@
+
+  return false
+
+end
```
Why all the empty lines?

anametiu replied:
I thought they will make the code easier to read. I removed them

pirvudoru commented on Problem3.rb:
```diff
@@ -0,0 +1,48 @@
+    while i != j do
+
+      if temp == array[j]
+        return true
```
Please keep a single return per function. Ruby, to help with this, returns the result of the last line in the method, without needing to specifically write `return`. Ex:

```
def sum(a,b)
  a + b
end

sum(1,3) # returns 4
```

anametiu replied:
I refactored all my code to take into consideration your remark

pirvudoru commented on Problem2.rb:
```diff
@@ -0,0 +1,32 @@
+
+    puts expect(find_max_numbers([4, 8, 2, 6, 9, 3])).to eq([9, 8])
+    puts expect(find_max_numbers([1, 2, 3, 4, 5, 6])).to eq([6, 5])
+    puts expect(find_max_numbers([3, 2, 7, 1, 1, 1])).to eq([7, 3])
```
Why do you have "puts" for every expect ?

anametiu replied:
Actually I looked more into testing and implemented unit tests in separate files using assert

pirvudoru commented on Problem2.rb:
```diff
@@ -0,0 +1,32 @@
+
+  end
+
+  return Array.new([first, second])
```
Just for reference: https://github.com/bbatsov/ruby-style-guide
Every major programming language has it's own.

anametiu replied:
Indeed, after reading the guide I got rid of the empty lines and unnecessary returns in all problems

pirvudoru commented on Problem3.rb:
```diff
@@ -0,0 +1,48 @@
+
+      j = temp - min
+
+    end
```
We write code for people, not machines. Spent ~ 15 minutes to figure out what the code does, but still a mistery to me. Can you please describe what you wanted to do ?

Btw, it is not 100% working

![image](https://user-images.githubusercontent.com/2707601/39755024-efb859b8-52cc-11e8-8fbe-8669bd3942c9.png)

anametiu replied:
The algorithm iterates over the array and moves each number into its corresponding index position. If the number is already found there, it means it's a duplicate. Indeed algorithm fails for array[3, 4, 1, 2]. I will correct it.

anametiu replied:
I completly reimplemented problem3 to meet the O(n) requirements

anametiu replied:
Algorithm fails because of the swap method.

pirvudoru commented:
Also, can check out: http://blog.cleancoder.com/uncle-bob/2014/12/17/TheCyclesOfTDD.html

anametiu committed:
Updated problem1 according to required changes

Implemented algorithm in one line without auxiliary variable
Removed empty lines
Removed return

anametiu committed:
Updated problem2.rb

Implemented for negative numbers
Rmeoved empty lines
Removed return

anametiu committed:
Uploaded tests for problem1 and problem2

Implemented test units using assert

anametiu committed:
Implemented problem3 again

For each index, take the value (absolute) and use it as index in the array. Make the value at the new index negative. If the value is already found as negative, this means the index was visited before, thus the value used as index was previously found in the array.

anametiu committed:
Update test_problem3.rb

alexandru-calinoiu commented:
@anametiu is really hard to review the code changes, please have only one solution for each problem. right now there are 2 files 'problem3' and 'Problem3'

alexandru-calinoiu commented:
Closing this due to lack of activity.

alexandru-calinoiu commented:
@anametiu I've reopened this following the email I've received from you.

anametiu committed:
Deleted old version of problem 3

anametiu committed:
Rename problem3.rb to Problem3.rb

alexandru-calinoiu commented:
Closing this because it to old..
